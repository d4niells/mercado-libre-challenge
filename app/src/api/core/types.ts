export type Product = {
  id: string;
  title: string;
  price: Price;
  picture: string;
  condition: string;
  free_shipping: boolean;
};

export type ProductDetail = Product & {
  description: string;
  sold_quantity: number;
};

export type Price = {
  currency: string;
  amount: number;
  decimals: number;
};

export type Author = {
  name: string;
  lastname: string;
};

export type SearchResponse = {
  author: Author;
  categories: string[];
  items: Product[];
};

export type GetProductResponse = {
  author: Author;
  item: ProductDetail;
};
