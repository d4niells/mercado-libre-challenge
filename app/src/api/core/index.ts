export * from './types';
export * from './mercado-livre';
