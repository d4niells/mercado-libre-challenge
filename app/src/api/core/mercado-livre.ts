import axios from 'axios';

import { SearchResponse } from '.';

export class MercadoLivre {
  static baseUrl = 'http://localhost:5000/api';

  static async searchProducts(query: string): Promise<SearchResponse> {
    const response = await axios.get<SearchResponse>(
      `${this.baseUrl}/items?q=${query}`
    );

    return response.data;
  }

  static async getProductById(productId: string): Promise<SearchResponse> {
    const response = await axios.get<SearchResponse>(
      `${this.baseUrl}/items/${productId}`
    );

    return response.data;
  }
}
