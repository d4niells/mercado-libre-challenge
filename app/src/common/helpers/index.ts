export { default as formatPrice } from './format-price';
