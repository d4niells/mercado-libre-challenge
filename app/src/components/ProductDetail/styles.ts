import styled, { css } from 'styled-components';

import ButtonBase from 'components/Common/Button';

export const Container = styled.div`
  ${({ theme }) => css`
    padding: ${theme.spacings.large};
  `}
`;

export const Wrapper = styled.div`
  display: grid;
  grid-template-columns: 70% 30%;
  grid-template-rows: 1;
  grid-template-areas: 'image payment';
`;

export const Image = styled.img`
  width: 680px;
  height: 460px;
  object-fit: contain;
  grid-area: image;
`;

export const PaymentInfo = styled.div`
  grid-area: payment;
  ${({ theme }) => css`
    h1 {
      font-size: ${({ theme }) => theme.font.sizes.xxxlarge};
      padding: ${({ theme }) => theme.spacings.large} 0;
      font-weight: 400;
    }

    h2 {
      font-size: ${theme.font.sizes.xlarge};
      font-weight: 600;
    }

    h3 {
      margin-bottom: ${theme.spacings.medium};
      font-size: ${theme.font.sizes.small};
      color: ${theme.colors.black};
      font-weight: 400;
    }
  `}
`;

export const Button = styled(ButtonBase)`
  ${({ theme }) => css`
    width: 100%;
    padding: ${theme.spacings.medium};
    font-weight: 400;
    font-size: ${theme.font.sizes.medium};
    border-radius: ${theme.border.radius.medium};

    background-color: ${theme.colors.secondary};
    color: ${theme.colors.white};
    transition: background-color 0.2s ease-in;

    &:hover {
      background-color: ${theme.colors.secondaryDark};
    }
  `}
`;

export const Description = styled.div`
  ${({ theme }) => css`
    h1 {
      margin: ${theme.spacings.large} 0;
      font-size: ${theme.font.sizes.xxlarge};
      color: ${theme.colors.black};
      font-weight: 400;
    }

    p {
      font-size: ${theme.font.sizes.medium};
      color: ${theme.colors.grey};
    }
  `}
`;
