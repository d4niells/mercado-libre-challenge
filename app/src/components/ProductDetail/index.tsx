import { ProductDetail as TProductDetail } from 'api/core';

import { formatPrice } from 'common/helpers';

import * as S from './styles';

type Props = {
  data: TProductDetail;
};

const ProductDetail = ({ data }: Props) => {
  const condition = data.condition === 'new' ? 'Nuevo' : 'Usado';
  const label = data.sold_quantity > 1 ? 'vendidos' : 'vendido';

  return (
    <S.Container>
      <S.Wrapper>
        <S.Image
          src={data.picture}
          alt="Image do produto no tamanho 680x460 ao lado das informações do produto"
        />
        <S.PaymentInfo>
          <h3>
            {condition} - {data.sold_quantity} {label}
          </h3>
          <h2>{data.title}</h2>
          <h1>{formatPrice(data.price)}</h1>
          <S.Button>Comprar</S.Button>
        </S.PaymentInfo>
      </S.Wrapper>
      <S.Description>
        <h1>Descripción del producto</h1>
        <p>{data.description}</p>
      </S.Description>
    </S.Container>
  );
};

export default ProductDetail;
