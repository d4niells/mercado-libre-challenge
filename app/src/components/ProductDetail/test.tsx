import { render } from 'common/tests';

import { ProductDetail as TProductDetail } from 'api/core';

import ProductDetail from '.';

const PRODUCT_BASE: TProductDetail = {
  id: 'ML',
  condition: 'new',
  free_shipping: true,
  title: 'Product title',
  description: 'Product description',
  picture: 'http://http2.mlstatic.com/D_976592-MLA46432963374_062021-I.jpg',
  price: {
    amount: 144,
    currency: 'ARS',
    decimals: 2
  },
  sold_quantity: 3
};

describe('<ProductDetail />', () => {
  it('should render product detail correctly', () => {
    const { container } = render(<ProductDetail data={PRODUCT_BASE} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
