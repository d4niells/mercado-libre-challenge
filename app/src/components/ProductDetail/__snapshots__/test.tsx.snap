// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetail /> should render product detail correctly 1`] = `
.c0 {
  display: grid;
  grid-template-columns: 70% 30%;
  grid-template-rows: 2;
  grid-template-areas: 'main main' 'description .';
  padding: 32px;
}

.c1 {
  grid-area: main;
  display: grid;
  grid-template-columns: 70% 30%;
  grid-template-rows: 1;
  grid-template-areas: 'image payment';
}

.c2 {
  width: 680px;
  height: 460px;
  object-fit: contain;
  grid-area: image;
}

.c3 {
  grid-area: payment;
}

.c3 h1 {
  font-size: 46px;
  padding: 32px 0;
  font-weight: 400;
}

.c3 h2 {
  font-size: 24px;
  font-weight: 600;
}

.c3 h3 {
  margin-bottom: 16px;
  font-size: 14px;
  color: #333333;
  font-weight: 400;
}

.c4 {
  width: 100%;
  padding: 16px;
  font-weight: 400;
  font-size: 16px;
  border-radius: 4px;
  background-color: #3483FA;
  color: #FFFFFF;
  -webkit-transition: background-color 0.2s ease-in;
  transition: background-color 0.2s ease-in;
}

.c4:hover {
  background-color: #2968c8;
}

.c5 {
  grid-area: description;
}

.c5 h1 {
  margin: 32px 0;
  font-size: 28px;
  color: #333333;
  font-weight: 400;
}

.c5 p {
  font-size: 16px;
  color: #999999;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      alt="Image do produto no tamanho 680x460 ao lado das informações do produto"
      class="c2"
      src="http://http2.mlstatic.com/D_976592-MLA46432963374_062021-I.jpg"
    />
    <div
      class="c3"
    >
      <h3>
        Nuevo
         - 
        3
         
        vendidos
      </h3>
      <h2>
        Product title
      </h2>
      <h1>
        $ 144,00
      </h1>
      <button
        class="c4"
      >
        Comprar
      </button>
    </div>
  </div>
  <div
    class="c5"
  >
    <h1>
      Descripción del producto
    </h1>
    <p>
      Product description
    </p>
  </div>
</div>
`;
